### PR TITLE
Refactor GUI tests

### PR DESCRIPTION
**Pull request by rafaabc**
@wlsf82 This PR is related to the refactoring of tests as we discussed.

Here are the main points:

- Extracted the base Url from the spec file to cypress.json
- Abstracted the login steps into custom commands to put it in a BeforeEach method
- Compiled the 3 test cases into one
- Created custom commands to the steps of create, edit and delete a note so they can be easily reused in other tests if necessary

### CODE REVIEW

File: cypress/fixtures/example.json [hunk 1]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (wlsf82) [inline cypress/fixtures/example.json]: Hi, @rafaabc,

Why did you add this fixture if it's not being used?

File: cypress/integration/crud.spec.js [hunk 2]
```diff
@@ -1,36 +1,16 @@
...
-    cy.get('.list-group:contains(My note updated)').should('be.visible')
-  })
-
-  it('deletes a note', () => {
-    cy.get('.list-group').contains('My note updated').click()
-    cy.contains('Delete').click()
-
-    cy.get('.list-group:contains(My note updated)').should('not.exist')
+describe('Notes', () => {
+  beforeEach(() => cy.login(Cypress.env('user'), Cypress.env('password')))
```
Comment (wlsf82) [inline cypress/integration/crud.spec.js]: What about just `cy.login()`, defaulting the user and password directly on the custom command?

File: cypress/fixtures/example.json (hunk 1 above)
Comment (rafaabc) [reply cypress/fixtures/example.json]: @wlsf82 It was created during the setup of Cypress and I'm going to ignore this file in a new commit.

File: cypress/integration/crud.spec.js (hunk 2 above)
Comment (rafaabc) [reply cypress/integration/crud.spec.js]: @wlsf82 Nice! Let me try and I'll update it.

File: cypress/support/commands.js [hunk 3]
```diff
@@ -23,3 +24,31 @@
 //
 // -- This will overwrite an existing command --
 // Cypress.Commands.overwrite('visit', (originalFn, url, options) => { ... })
+Cypress.Commands.add('login', () => {
```
Comment (wlsf82) [inline cypress/support/commands.js]: When using the `cy.login()` function in the test file, you're allowing the user of this command to pass the user and password, but here you're not accepting those arguments.
I suggest that you accept them (maybe a user object), but default the values to the info coming from the `cypress.env.json` file. Something like this:

```js
Cypress.Commands.add('login', (user = {
  email: Cypress.env('user'),
  password: Cypress.env('password')
}) => {
  cy.visit('/login')
  cy.get('#email')
    .should('be.visible')
    .type(user.email)
  cy.get('#password')
    .should('be.visible')
    .type(user.password)
  cy.get('button[type="submit"]')
    .should('be.visible')
    .click()
})
```

File: cypress/support/commands.js [hunk 4]
```diff
@@ -23,3 +24,31 @@
...
+  cy.visit('/');
+  cy.get('.navbar-nav a:contains(Login)').click();
+  cy.get('#email').type(Cypress.env('user'));
+  cy.get('#password').type(Cypress.env('password'));
+  cy.get('button[type="submit"]').click();
+})
+
+Cypress.Commands.add('createsANote', data => {
+  cy.contains('Create a new note', { timeout: 30000 }).should('be.visible').click();
```
Comment (wlsf82) [inline cypress/support/commands.js]: Why are you overriding the default timeout here?

File: cypress/support/commands.js [hunk 5]
```diff
@@ -23,3 +24,31 @@
...
+  cy.get('#content').type(data.content);
```
Comment (wlsf82) [inline cypress/support/commands.js]:
```suggestion
  cy.get('#content').should('be.visible').type(data.content);
```

File: cypress/support/commands.js [hunk 6]
```diff
@@ -23,3 +24,31 @@
...
+  cy.contains('Create').click();
```
Comment (wlsf82) [inline cypress/support/commands.js]:
```suggestion
  cy.contains('Create').should('be.visible').click();
```

File: cypress/support/commands.js [hunk 7]
```diff
@@ -23,3 +24,31 @@
...
+  cy.get('.list-group').should('contain', data.content);
+})
+
+Cypress.Commands.add('editsANote', data => {
+  cy.get('.list-group').contains(data.content).click();
```
Comment (wlsf82) [inline cypress/support/commands.js]:
```suggestion
  cy.get('.list-group').contains(data.content).should('be.visible').click();
```

Comment (wlsf82) [reply cypress/support/commands.js]: I suggest that you "always" check that the element is visible before interacting with it to make the tests more robust and close to how a real user uses a web app.

File: cypress/support/commands.js [hunk 8]
```diff
@@ -23,3 +24,31 @@
...
+  cy.get('#content').clear().type(data.newContent);
+  cy.contains('Save').click();
```
Comment (wlsf82) [inline cypress/support/commands.js]: For this project, I was not using semi-colons at the end. I noticed that you're using it in some places and not in others.

Let's have a standard (I suggest without it) and use it everywhere.

File: cypress/support/commands.js [hunk 9]
```diff
@@ -23,3 +24,31 @@
...
+  cy.get('.list-group').should('contain', data.newContent);
+  cy.get(`.list-group:contains(${data.newContent})`).should('be.visible');
+})
+
+Cypress.Commands.add('deletesANote', data => {
+  cy.get('.list-group').contains(data.newContent).click()
+  cy.contains('Delete').click()
+  cy.get(`.list-group:contains(${data.newContent})`).should('not.exist')
+})
```
Comment (wlsf82) [inline cypress/support/commands.js]: It's good practice to put an empty line at the end of the file. This helps in rendering the code in different IDEs

File: package.json [hunk 10]
```diff
@@ -10,5 +10,15 @@
   "license": "MIT",
   "devDependencies": {
     "cypress": "^6.8.0"
-  }
+  },
+  "main": "index.js",
```
Comment (wlsf82) [inline package.json:14]: There's no `index.js` file in this project.

File: package.json [hunk 11]
```diff
@@ -10,5 +10,15 @@
...
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/rafaabc/tech-interview-sample-project.git"
```
Comment (wlsf82) [inline package.json:17]: This and the below URLs should point to the original repo, not the fork.

**Review by wlsf82 — CHANGES_REQUESTED** [review]
Hi, @rafaabc,
Thanks for the PR.
I've left a few comments and suggestions.
Great work! 🚀

File: cypress/fixtures/example.json (hunk 1 above)
Comment (wlsf82) [reply cypress/fixtures/example.json]: Do not put it on `.gitignore` if that's what you meant. I'd simply delete it.

Comment (rafaabc) [reply cypress/fixtures/example.json]: It makes sense. I'll delete it as it won't be used

Comment (rafaabc) [reply cypress/fixtures/example.json]: Done!

File: cypress/support/commands.js (hunk 3 above)
Comment (rafaabc) [reply cypress/support/commands.js]: Done!

File: cypress/integration/crud.spec.js (hunk 2 above)
Comment (rafaabc) [reply cypress/integration/crud.spec.js]: Done!

File: cypress/support/commands.js (hunk 4 above)
Comment (rafaabc) [reply cypress/support/commands.js]: After the application fills the login form and clicks on the submit button, it's been taking more than the default time of 4000ms and I get a Timeout error before I get redirected to the main page. Meanwhile, I figured out and added an assertion that waited for the new page to be loaded. I'll update in a new commit

File: cypress/support/commands.js (hunk 5 above)
Comment (rafaabc) [reply cypress/support/commands.js]: Done!

File: cypress/support/commands.js (hunk 6 above)
Comment (rafaabc) [reply cypress/support/commands.js]: Done!

File: cypress/support/commands.js (hunk 7 above)
Comment (rafaabc) [reply cypress/support/commands.js]: Done!

File: cypress/support/commands.js (hunk 9 above)
Comment (rafaabc) [reply cypress/support/commands.js]: Very interesting! I didn't know about that.

Done!

File: cypress/support/commands.js (hunk 8 above)
Comment (rafaabc) [reply cypress/support/commands.js]: Done!

I removed all of them.

File: cypress/integration/crud.spec.js [hunk 12]
```diff
@@ -1,36 +1,16 @@
...
+describe('Notes', () => {
+  beforeEach(() => cy.login())
+  
+  it('Creates, edits and deletes a note', () => {    
+    const data = {
+      content: 'My note',
+      newContent: 'My note updated'
+    }
+    
+    cy.createsANote(data)
```
Comment (wlsf82) [inline cypress/integration/crud.spec.js:14]: Ideally, expectations should be explicit on the test body to understand exactly what's the behavior we're testing.

File: cypress/integration/crud.spec.js [hunk 13]
```diff
@@ -1,36 +1,16 @@
...
+    cy.editsANote(data)
+    cy.deletesANote(data)
   })
-})
+})
```
Comment (wlsf82) [inline cypress/integration/crud.spec.js]: It's not space. It's an empty line.

File: cypress/support/commands.js [hunk 14]
```diff
@@ -23,3 +24,34 @@
...
 // Cypress.Commands.overwrite('visit', (originalFn, url, options) => { ... })
+Cypress.Commands.add('login', (user = {
+  email: Cypress.env('email'),
+  password: Cypress.env('password')
+}) => {
+  cy.visit('/login');
+  cy.get('#email').should('be.visible').type(user.email)
+  cy.get('#password').should('be.visible').type(user.password)
+  cy.get('button[type="submit"]').should('be.visible').click()
+  cy.url().should('eq', 'https://notes-serverless-app.com/')
```
Comment (wlsf82) [inline cypress/support/commands.js]: If I pass invalid credentials, this step will always fail. Maybe the expectation should go in the test, not the app action.

**Review by wlsf82 — CHANGES_REQUESTED** [review]
Hi, @rafaabc,
I've left a few more comments.
Feel free to ask for my review again after you're done with them.

File: cypress/support/commands.js [hunk 15]
```diff
@@ -23,3 +24,34 @@
...
+  cy.get('#content').should('be.visible').clear().type(data.newContent)
+  cy.contains('Save').should('be.visible').click()
+  cy.get('.list-group').should('contain', data.newContent)
+  cy.get(`.list-group:contains(${data.newContent})`).should('be.visible')
+})
+
+Cypress.Commands.add('deletesANote', data => {
+  cy.get('.list-group').contains(data.newContent).click()
+  cy.contains('Delete').click()
+  cy.get(`.list-group:contains(${data.newContent})`).should('not.exist')
```
Comment (wlsf82) [inline cypress/support/commands.js]: Ideally, we should wait for the redirect to happen before running the assertion. Also, it would be better if the assertion was on the test itself, not the custom command.

File: cypress/integration/crud.spec.js (hunk 12 above)
Comment (rafaabc) [reply cypress/integration/crud.spec.js:14]: Done!

File: cypress/integration/crud.spec.js (hunk 13 above)
Comment (rafaabc) [reply cypress/integration/crud.spec.js]: Done!

File: cypress/support/commands.js (hunk 14 above)
Comment (rafaabc) [reply cypress/support/commands.js]: Done!

File: cypress/support/commands.js (hunk 15 above)
Comment (rafaabc) [reply cypress/support/commands.js]: Done! I moved all of the assertions from the custom commands to the test body.

**Comment by rafaabc** [comment]
@wlsf82 I'm jsut having some issues running on headless mode because of timeout. Login page is taking sometimes more than 4000ms so I'm trying to create an api_login custom command for instance so it could be solved properly :)

**Comment by wlsf82** [comment]
> @wlsf82 I'm jsut having some issues running on headless mode because of timeout. Login page is taking sometimes more than 4000ms so I'm trying to create an api_login custom command for instance so it could be solved properly :)

What about just increasing the timeout? In the end, this is just a sample project.

**Comment by rafaabc** [comment]
> > @wlsf82 I'm jsut having some issues running on headless mode because of timeout. Login page is taking sometimes more than 4000ms so I'm trying to create an api_login custom command for instance so it could be solved properly :)
> 
> What about just increasing the timeout? In the end, this is just a sample project.

Done! And it was a good way to get familiar with Timeout concepts reading the documentation: https://docs.cypress.io/guides/references/configuration#Timeouts

**Comment by wlsf82** [comment]
I'm closing it without merging since it was just an exercise.
